oeps/oep-0028-content-theming-in-react.rst

```diff
@@ -0,0 +1,246 @@
+
+
+Now if we want to customize our _Header component, we can easily do it like
+
+.. code-block:: js
+
+    // SiteLogo being updated in Header
+    class MyThemedHeader extends _Header {
+        render() {
+            return (
+                <header>
+                    {/* Replace <SiteLogo/> with a fancy widget */}
+                    <MyCustomAnimatedLogoWidget/>
+                    <MainNav/>
+                    <UserAvatar/>
+                </header>
+            );
+        }
+    }
+    // Custom theme:
+    export const Header = MyThemedHeader;
+    export const MainNavWrapper = _MainNavWrapper;
+
+
+* We will provide support to add data or other components by using `props.children`. The `props.children` property is provided from React and can be extended to include specialized children like left and right for a Pane component. An example of this can be
+
+.. code-block:: js
+
+    const Header = (props) => {
+        return (
+            <h1>{props.title}</h1>
+            {props.children}
+        )
+    }
+
+    // use
+    <Header title={"Open Edx"} >
+        <h4>Open Source MOOC platform</h4>
+    </Header>
+
+Another example where specialized children are passed
+
+.. code-block:: js
+
+    const Pane = (props) => {
+        return (
+            <div className="pane">
+                <div className="pane-left">
+                    {props.left}
+                </div>
+                <div className="pane-right">
+                    {props.right}
+                </div>
+            </div>
+
+        )
+    }
+
+    // use
+    <Pane left={<Image />} right={<Explanation />} />
+
+
+* We will use function and placeholder to add additional content in customizable components. We can take an example of the above DefaultTheme and see _MainNav where it has support to add additional nav links by overriding `extraNavLinks` function.
+
+.. code-block:: js
+
+    // Customizable Main Navigation Area
+    class MyThemedNav extends _MainNav {
+        get extraNavLinks() {
+            return [
+                <a href="/about">About Us</a>,
+            ];
+        }
+    }
+
+    // Custom theme:
+    export const MainNav = MyThemedNav;
+    export const MainNavWrapper = _MainNavWrapper;
+
+
+* We will provide support for a global redux store which will be Open edX global redux store. This will act as a central place to store data and will have access to all data available to the system.
+
+* We will consider the layout of the data in the Open edX global redux store as a stable API. We will provide support to pre-fill the store with some common data like current user, current course, list of courses enrolled, etc. We will provide the choice for themes to fetch data that's not part of the Open edX global redux store from REST API's using built-in redux actions and store it in their own separate redux store. We will announce breaking changes if the layout of the data changes in global store.
```

Comment:
The redux store would be specific to each frontend (LMS, Studio, ecommerce, etc.) and not global to all of "Open edX".

> using built-in redux actions and store it in their own separate redux store

Actions tend to be fairly coupled to the reducers and overall store state, so I'm not sure how useful it would be to provide standard action [creator]s but expect them to work with custom stores/reducers. It might be better to just provide a standard API client library for the REST APIs, without redux actions, and let each custom theme that needs to use them include its own redux actions/reducers/state. But maybe this would help reduce boilerplate etc., and most any reducer could be adapter to ingest standard actions. Curious to know what others think.

Reply:
I don't think we can expect that there will be 100% coverage for each API in the frontend, so people will have to create their own actions. I don't think it's reasonable to have the actions and reducers etc for data not used by the default theme. Where a built-in action exists, chances are that data is already loaded, and if not, the standard actions can perhaps be used.

While an API client could be automatically generated using Swagger or such, I think there should probably be multiple API clients for multiple sets of APIs because the API is huge so why should all that code be loaded if most of it isn't used. If a custom theme uses some API that the default theme doesn't, it can also include the relevant API client. This probably should cover the whole API.

Reply:
@xitij2000 @bradenmacdonald I have updated this point but I am not sure if it covers your concern. Please let me know your thoughts on this. Thanks.